### DISCUSSION TITLE
Color bar of mesh plot doesn't match the plot

### DISCUSSION THREAD

**Discussion by derne058** [pyvista/pyvista#6634] [Q&A]
### First check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the PyVista documentation, with the integrated search.
- [X] I already searched in Google "How to X in PyVista" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I will NOT make duplicate discussions or issues in the future.

### Commit to Help

- [x] I commit to help with one of those options.

### Sample Code What is the problem, question, or error?
Write a short description telling me what you are doing, what you expect to happen, and what is currently happening.

I'm struggling with matching a color bar of a mesh plot. I'm using pyvista and I'm plotting an array that contains scalar values that range from 0 to 1. Furthermore, each value of my array represent the "porosity" of each element (the 1st value in the array is linked to the 1st element, the 2nd value in the array is linked to the 2nd element, and so on).
Here is the code:
```python
# plot element density:
mesh = mapdl.mesh.grid    # gets the mesh from ANSYS MAPDL object (mesh.grid create a mesh that PyVista is able to plot)
mesh.cell_data["Element density"] = x_full
plot = pv.Plotter()
plot.add_mesh(mesh, scalars="Element density", show_edges=True, cmap="coolwarm", clim=[rho_min, 1], show_scalar_bar=True)
plot.set_background("white") 
plot.show()"

I already tried to share my problem on PyMapdl community, but we were not able to fix it.
```

### Description

- The code takes a mesh and an array of values ranging from 0 and 1 (each element in the array is associated with specific mesh element)
- The output is a mesh plot
- The output returns a color bar that doesn't match the mesh element colors

### System Information

```shell
OS : Windows
            CPU(s) : 8
           Machine : AMD64
      Architecture : 64bit
               RAM : 15.7 GiB
       Environment : Jupyter
        GPU Vendor : Intel
      GPU Renderer : Intel(R) Iris(R) Plus Graphics
       GPU Version : 4.5.0 - Build 27.20.100.8681
  MathText Support : False

  Python 3.9.0 (tags/v3.9.0:9cf6752, Oct  5 2020, 15:34:40) [MSC v.1927 64 bit
  (AMD64)]

           pyvista : 0.42.3
               vtk : 9.2.6
             numpy : 1.26.0
        matplotlib : 3.8.0
            scooby : 0.7.4
             pooch : v1.7.0
            pillow : 10.0.1
           IPython : 8.17.2
             scipy : 1.11.4
              tqdm : 4.66.1
      nest_asyncio : 1.5.8
```

### Screenshots

![output](https://github.com/user-attachments/assets/7fb8e5d3-9cf0-4b5f-be52-ed6566812ac4)

**Comment by banesullivan** [comment]
Would you please try to reproduce this error with a [minimal, reproducible example](https://stackoverflow.com/help/minimal-reproducible-example)? This is definitely not normal or expected and I don't think we've seen this before. Without a fully reproducible example we probably cannot address this. Otherwise, I'm concerned this may be a specific hardware/GL issue on your system

**Reply by derne058** [reply]
Hi @banesullivan , thanks for replying!
I tried the following code:

```python
from ansys.mapdl.core import launch_mapdl
import pyvista as pv
import numpy as np
import math
import os

path = os.getcwd()
working_directory = "my working directory"

mapdl = launch_mapdl(run_location=working_directory, port=50056, override=True)
print(mapdl)

mapdl.clear()
mapdl.prep7()

mapdl.et(1, 185)
mapdl.mp('EX', 1, 1E7)

mapdl.block(0, 3, 0, 6, 0, 3)

mapdl.esize(0.25)
mapdl.vsweep(1)

edens_list = []
for elem in mapdl.mesh.enum:
    edens_list.append(math.sin(mapdl.queries.centrx(elem))*mapdl.queries.centry(elem))
edens = np.asarray(edens_list)
mesh = mapdl.mesh.grid
mesh["EDENS"] = edens
rho_min = min(edens_list)
print(rho_min)
plot = pv.Plotter()
plot.add_mesh(mesh=mesh, scalars='EDENS', show_edges=True, cmap="coolwarm", clim=[rho_min, 1],
              show_scalar_bar=True, n_colors=24)
plot.set_background("white") 
plot.show()
mapdl.exit()
```

And I got this plot:
![result](https://github.com/user-attachments/assets/5883a74b-10f4-4e66-a348-457c6d7377ae)

**Reply by banesullivan** [reply]
Hm, I don't think I can run MAPDL without a license? Would you please try to save the `mesh` to a VTK file for me: `mesh.save("mesh.vtk")`

**Reply by banesullivan** [reply]
Also, have you raised this issue with the mapdl developers? https://github.com/ansys/pymapdl

**Reply by derne058** [reply]
I'm sorry but I have some problems uploading the .vtk mesh file, but I have modified the script using the one found in the Pyvista documentation:

```py
from __future__ import annotations
import numpy as np
import pyvista as pv

values = np.linspace(0, 10, 1000).reshape((20, 5, 10))
values.shape

grid = pv.ImageData()

grid.dimensions = np.array(values.shape) + 1

grid.origin = (100, 33, 55.6)  # The bottom left corner of the data set
grid.spacing = (1, 5, 2)  # These are the cell sizes along each axis

grid.cell_data["values"] = values.flatten(order="F")  # Flatten the array

grid.plot(show_edges=True)

values = np.linspace(0, 10, 1000).reshape((20, 5, 10))
values.shape

grid = pv.ImageData()

grid.dimensions = values.shape

grid.origin = (100, 33, 55.6)  # The bottom left corner of the data set
grid.spacing = (1, 5, 2)  # These are the cell sizes along each axis

grid.point_data["values"] = values.flatten(order="F")  # Flatten the array

grid.plot(show_edges=True)
```

Running the code I got this:

![image](https://github.com/user-attachments/assets/a190a766-37fe-4ee4-a791-b0c7f2b79695)

**Reply by banesullivan** [reply]
I'm almost certain this is a graphics driver issue on Windows (I've seen similar issues when running on Windows virtual machines). Unfortunately, I don't know if there is much that can be done here, but I recommend checking and updating any graphics drivers.

The best bet is to create the same scalar bar with matplotlib and plot separately or overlay as a "Chart" -- I'm hoping to add an interface that will do this for you, so stay tuned

**Reply by derne058** [reply]
Thanks a lot for all your suggestions! In the meantime I'll update the graphics driver and I'll create a separate scalar bar using matplotlib.

**Reply by JanaBrunatova** [reply]
Does the following code work for you? I think that using pyvista Plotter might do the job. 
I am also a Windows user and I always work with off_screen rendering, it's just safer. :)

```py
import numpy as np
import pyvista as pv

pv.start_xvfb()

values = np.linspace(0, 10, 1000).reshape((20, 5, 10))

grid = pv.ImageData()

grid.dimensions = (np.array(values.shape) + 1) 

grid.origin = (100, 33, 55.6)  # The bottom left corner of the data set
grid.spacing = (1, 5, 2)  # These are the cell sizes along each axis

grid.cell_data["values"] = values.flatten(order="F") # Flatten the array

plotter = pv.Plotter(off_screen=True)
plotter.add_mesh(grid, show_edges=True)
plotter.screenshot("cell_data_grid.png")  # Save the off-screen plot to an image file
plotter.close()

values = np.linspace(0, 10, 1000).reshape((20, 5, 10))

grid = pv.ImageData()

grid.dimensions = values.shape

grid.origin = (100, 33, 55.6)  # The bottom left corner of the data set
grid.spacing = (1, 5, 2)  # These are the cell sizes along each axis

grid.point_data["values"] = values.flatten(order="F") # Flatten the array

plotter = pv.Plotter(off_screen=True) # Create a new PyVista plotter for off-screen plotting
plotter.add_mesh(grid, show_edges=True)
plotter.screenshot("point_data_grid.png")  # Save the off-screen plot to an image file
plotter.close()
```
![cell_data_grid](https://github.com/user-attachments/assets/55acf0fa-6310-4c92-9d16-ce50c6add5a8)

**Comment by JanaBrunatova** [comment]
Alternatively, you can save your colorbar separately using matplotlib and subsequently overlay with your plot using `plotter.add_logo_widget`. I hope it helps!

```py
import matplotlib.colors as mcolors
import matplotlib.pyplot as plt
import numpy as np
import pyvista as pv

pv.start_xvfb()

values = np.linspace(0, 10, 1000).reshape((20, 5, 10))

grid = pv.ImageData()

grid.dimensions = np.array(values.shape) + 1

grid.origin = (100, 33, 55.6)  # The bottom left corner of the data set
grid.spacing = (1, 5, 2)  # These are the cell sizes along each axis

grid.cell_data["values"] = values.flatten(order="F")  # Flatten the array in Fortran order

plotter = pv.Plotter(off_screen=True)  # Create a PyVista plotter for off-screen plotting
plotter.add_mesh(grid, show_edges=True, show_scalar_bar=False)


fig, ax = plt.subplots(figsize=(6, 0.5))  # Create a colorbar
fig.subplots_adjust(bottom=0.5)
cmap = plt.get_cmap("viridis")  # Set the colormap and norm
norm = mcolors.Normalize(vmin=0, vmax=10)  # Set the limits

cb = plt.colorbar(
    plt.cm.ScalarMappable(norm=norm, cmap=cmap),
    cax=ax,
    orientation="horizontal",
)  # Create a colorbar

cb.set_label("values", fontsize=14, labelpad=15)
cb.ax.xaxis.set_label_position("top")

cb.ax.tick_params(
    which="major",  # Major ticks
    direction="inout",  # Make ticks go inside and outside the colorbar
    length=10,  # Increase the length of the major ticks
    width=1,  # Adjust the width of the major ticks
)  # Customize the major ticks: make them appear both inside and outside, and increase their length

cb.ax.tick_params(labelsize=12)  # Customize the size of the tick labels

plt.savefig(
    "colorbar_standalone.png", dpi=150, bbox_inches="tight", transparent=True
)  # Save the colorbar figure

plotter.add_logo_widget(logo="colorbar_standalone.png", position=(0.4, 0), size=(0.6, 0.2))
plotter.screenshot("cell_data_grid.png")  # Save the off-screen plot to an image file
plotter.close()
```
![cell_data_grid](https://github.com/user-attachments/assets/cc108023-5ee8-4aa0-973e-0fd62d3f0cf7)

**Reply by derne058** [reply]
Hi @JanaBrunatova, thanks a lot! The first code doesn't work for me but the second one is working well!
